src/string.cpp

```diff
@@ -0,0 +1,598 @@
+#include <algorithm>
+#include <array>
+#include <cstddef>
+#include <ios>
+#include <iostream>
+#include <memory>
+#include <optional>
+#include <regex>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+#include <string_view>
+#include <unordered_map>
+#include <variant>
+#include <vector>
+
+#include "MiniLua/source_change.hpp"
+#include "MiniLua/string.hpp"
+#include "MiniLua/utils.hpp"
+#include "MiniLua/values.hpp"
+
+namespace minilua {
+template <class Result>
+auto static try_value_as(
+    Value s, const std::string& method_name, int arg_index, bool needs_int_representation = false)
+    -> Result {
+    if (s.is_number()) {
+        try {
+            return std::get<Number>(s).convert_to<Result>(needs_int_representation);
+        } catch (const std::runtime_error& e) {
+            throw std::runtime_error(
+                "bad argument #" + std::to_string(arg_index) + " to '" + method_name +
+                "' (number has no integer representation)");
+        }
+    }
+
+    auto tmp = Number(1);
+    try {
+        if (s.is_string()) {
+            Value v = s.to_number();
+            if (v == Nil()) {
+                throw std::runtime_error("");
+            }
+            tmp = std::get<Number>(v);
+        } else {
+            throw std::runtime_error("");
+        }
+    } catch (const std::runtime_error& e) {
+        throw std::runtime_error(
+            "bad argument #" + std::to_string(arg_index) + " to '" + method_name +
+            "' (number expected, got " + s.type() + ")");
+    }
+
+    try {
+        return tmp.convert_to<Result>(needs_int_representation);
+    } catch (const std::runtime_error& e) {
+        throw std::runtime_error(
+            "bad argument #" + std::to_string(arg_index) + " to '" + method_name +
+            "' (number has no integer representation)");
+    }
+}
+
+auto static try_value_is_string(const Value& val, const std::string& method_name, int arg_index)
+    -> Value {
+    if (val.is_string() || val.is_number()) {
+        return val.to_string();
+    }
+    throw std::runtime_error(
+        "bad argument #" + std::to_string(arg_index) + " to '" + method_name +
+        "' (string expected, got " + val.type() + ")");
+}
+
+auto static parse_string(const std::string& str, std::vector<Value> args) -> std::string {
+    std::string_view s(str);
+    std::stringstream ss;
+
+    // find escapes and replace them with the argument
+    int numEscapes = 0;
+    size_t pos = 0;
+    while (pos < s.length()) {
+
+        size_t start_pos = s.find('%', pos);
+        if (start_pos == std::string_view::npos) {
+            ss << s.substr(pos);
+            break;
```

Comment:
is this not just `return s;` ?

Reply:
No. Because this returns all text after the escape. for example: "hello %s world", then this returns " world" so that in the string stream, there is the whole formated text.

Your thesis is only true in the case when there is no escape string remaining.